xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/adduorg.vm

```diff
@@ -17,22 +17,49 @@
 ## Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 ## 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 ## ---------------------------------------------------------------------------
+##
+## Checks if this is valid to add the addedGroup to the currentGroup.
+## This verification is based on three criteria:
+## - the current and added groups are not the same group
+## - the current group does not already have the added group in its sub-groups
+## - this current group is not sub-group of the added group
+##
+## The boolean result is set on the $accept variable.
+##
+#macro (checkGroups $currentGroup $addedGroup)
+  #set ($members = $services.user.group.getMembers($currentGroup, true))
+  #set ($membersOf = $services.user.group.getGroupsFromAllWikis($currentGroup))
+  #set ($currentGroupDR = $services.model.resolveDocument($currentGroup))
```

Comment:
You should move the resolves before the calls to getMembers and getGroupsFromAllWikis so that you don't end up doing them twice (once automatically and once explicitly).